### ISSUE TITLE
Enabling/disabling and displaying high resolution audio

### ISSUE THREAD

**Issue by JeanGoulet** [acidhax/chromecast-audio-stream#19]
Glad to have found audio-cast.exe. Run it and hear your audio. Simple and straighforward.

My question is to know if audio-cast.exe could 1) view the current setting for "High dynamic range", and 2) enable/disable it.

Right now audio-cast.exe is the only software I have found which sends audio files to the Chromecast Audio (CCA) device, and to the Windows default audio playback device (PC speakers), at the same time. Great for debugging audio quality.

I was searching for this because my music files are stored on the PC's hard drive, and I usually play that local music with Foobar2000 or MusicBee.

The primary failure I am getting is that the Google "Chromecast App" often fails to maintain the checkmark for "Enable High Dynamic Range" (for HiFi systems). It seems rebooting the CCA causes the checkmark to be lost every time. I have no proof so far that it even works.

Hence my request for audio-cast.exe to come to the rescue and overcome that weakness with an option (command line or GUI).

I have a fully digital audio setup since I am using the CCA's optical digital output. My preference is to play lossless FLAC format. That's the main reason why I am not interested in the cloud-based "Google Play Music" where they (not us) control the quality level they feel is right for feeding audio. Thinking of how bad Sirius/XM satellite radio has become with sub-par bitrates.

For those who aren't aware:
- Inside the 3.5mm jack lies an optical transmitter/receiver that allows Toslink to be connected via any 3.5mm - Toslink adapter. Then you plug an optical cable from the CCA to your HiFi receiver. A red light comes out of the cable.
- The official specs for the CCA state that it is capable of High-Resolution Audio: 96KHz/24-bit (i.e. about 4.4 Mbps for stereo).

Otherwise, I am guessing that the CCA should be playing at 44.1 kHz/16-bit ("CD quality") per channel (i.e. about 1.4 Mbps for stereo) so I investigated the Wifi stream with an Asus RT-AC68U router and saw the traffic monitor showing around 20 kbytes/sec i.e 160 kbps which cannot be lossless audio. This is on page Adaptive QOS - Traffic Monitor - Wireless (2.4 GHz)

I sure would like audio-cast.exe to tell us what the CCA is actually doing.

My setup is Windows 10, 64-bit on up to date hardware and I should note that audio-cast.exe 32-bit has crashed a few times and the icon in the taskbar is dim (instead of the expected brightness).

I am surprised that there is no 64-bit version, perhaps it would help solve that? Only found a 32-bit version on the Download page: https://github.com/acidhax/chromecast-audio-stream/releases/download/0.2.5/audio-cast-1.0.0-win32-ia32.zip

A big thanks for coming up with this program.

Jean
Montreal

**Comment by matbeedotcom** [comment]
You are very correct for figuring everything out.

I do not, in fact, send "high resolution audio". This is most definitely on my to-do list. I'll most likely just get FLAC working, and ship it to the Chromecast. I'm just converting to mp3, and using whatever the default audio quality is in FFmpeg.

I will figure this out, relatively soon :)

**Comment by JeanGoulet** [comment]
Ok that's great to hear

By the way I also monitored Performance - Ethernet in Win10 Task Manager and when casting is activated, the rate goes from 0 to 160 kbps (thus matching what the router reports).

**Comment by pandathumb** [comment]
any chance you think you will be able to update this to cast full resolution audio? i love the program, and definitely wish it didn't send a low bitrate to my system.

**Comment by matbeedotcom** [comment]
So I've done some research and the plugin I'm using to capture system audio is limited to "s16le           PCM signed 16-bit little-endian". So I'm updating the app with uncompressed audio, but I highly doubt 7.1, etc will happen.

**Comment by JeanGoulet** [comment]
All right, thanks for that news update. Hopefully once it's compiled and run, the 16-bit PCM mode will generate 1.4 Mbps of stereo audio, and not the current 160 kbps. Of course 24-bit PCM at 96 KHz would still be nice to attain, but realistically, it would not likely provide a discernable difference except for minor subtleties that could only be heard in the most perfect, quiet environment.

**Comment by matbeedotcom** [comment]
From my tests I'm still only getting about 200-300kbps which I found strange, but I'll let you figure that out for me :)

**Comment by cabasse** [comment]
is this all running in node, and configurable vs editing lib.js/main.js? i found `getFFmpegCommandOSX = function getFFmpegCommandOSX(soundflowerDevice) {
    var command = (0, _fluentFfmpeg2.default)();
    command.setFfmpegPath(ffmpegPath);
    command.input('none:' + soundflowerDevice.index).inputFormat('avfoundation').outputFormat("adts").outputOptions(["-strict -2", "-b:a 320k"]).on('start', function (commandLine) {
        console.log('Spawned Ffmpeg with command: ' + commandLine);
    }).on('error', function (err, one, two) {
        console.log('An error occurred: ' + err.message);
        console.log(two);
    }).on('end', function () {
        console.log("end");
    });`
(and changed the default quality from 192k to 320k) but is this being re-read into the running application when i start audio-cast.exe or is that precompiled, and you just included the the various files that are a part of it in the folder?

**Comment by Pealy** [comment]
Have you tried writing something to the console to see if it gets written? That should tell you.

**Comment by cabasse** [comment]
I don't have visibility into the console (That I'm aware of at least) when I run audio cast.exe

**Comment by cabasse** [comment]
But I do see what you're getting at. I'll try modifying the file with some additions to write a logfile if the code fired

**Comment by Pealy** [comment]
I tried it and couldn't see my change reflected in the console so I'm pretty sure it's compiled.  There's a great guide available but I never got enough time to try it
https://github.com/nwjs/nw.js/wiki/how-to-package-and-distribute-your-apps

**Comment by mcprat** [comment]
Hi all. Decided to randomly join in.

@JeanGoulet CCA has a combination jack right? so can't you just plug in a toslink cable without any adapter? Also if you have an android phone, I have found the Chrome Cast app on android to be very reliable with the CCA settings. I have also noticed issues with settings with the desktop app.

@acidhax @luigiplr I second that the app should have settings for bitrate and depth. Along with that, the ability to choose between which sound driver is captured like WASAPI or DirectSound or otherwise would be very very useful to me.

**Comment by JeanGoulet** [comment]
@mpratt14 Yes the Chromecast Audio has a combination jack that accepts both:
a) 3.5mm stereo plug (analog)
b) miniToslink (digital) and I use an adapter ($1) for miniToslink to regular Toslink.

and I confirmed that both of them work. I control streaming to Chromecast Audio mostly from a Win10 PC (not Android).

**Comment by mcprat** [comment]
Ah... I did not realize miniToslink was a thing....

I've gotten frustrated enough with all of the things I have tried to get system wide audio over to the chromecast. I don't really think anyone's software is developed enough yet. I have found ways to be satisfied with controlling my CCA with android. I mean, even the apps give me trouble, lol. Best way to use it on android btw is to mirror with the Google Cast app.

What if you controlled the audio output with android but it was actually pushing your Win10 audio? This would be similar to BubbleUPnP but would be less limited I guess.

Also if you are using Toslink connection then the "Dynamic Range" setting is irrelevant.

> "When Chromecast Audio is connected through an optical interface (see here for compatible cables), it will always use Full Dynamic Range. Setting Full Dynamic Range to OFF won't have an effect as it will remain ON."

From: https://support.google.com/chromecast/answer/6290498
